File: Lecture5Code/BGCD.pl
#!/usr/bin/perl -w 

##  BGCD.pl

use strict;

die "\nUsage:   $0  <integer>  <integer>\n" unless @ARGV == 2;

my ($a,$b) = @ARGV;

my $gcdval = bgcd($a,$b );
print "\nBGCD: $gcdval\n\n";

sub bgcd {
    my ($a,$b) = @_;
    return $a if $a == $b;                                      #(A)
    return $b if $a == 0;                                       #(B)
    return $a if $b == 0;                                       #(C)
    if (~$a & 1) {                                              #(D)
        if ($b & 1) {                                           #(E)
            return bgcd($a >> 1, $b);                           #(F)
        } else {                                                #(G)
            return bgcd($a >> 1, $b >> 1) << 1;                 #(H)
        }
    }
    return bgcd($a,$ b >> 1) if (~$b & 1);                      #(I)
    return bgcd( ($a - $b) >> 1, $b) if ($a > $b);              #(J)
    return bgcd( ($b - $a) >> 1, $a );                          #(K)
}

File: Lecture5Code/BGCD.py
#!/usr/bin/env python

##  BGCD.py

import sys
if len(sys.argv) != 3:
    sys.exit("\nUsage:   %s  <integer>  <integer>\n" % sys.argv[0])

a,b = int(sys.argv[1]),int(sys.argv[2])

def bgcd(a,b):
    if a == b: return a                                         #(A)
    if a == 0: return b                                         #(B)
    if b == 0: return a                                         #(C)
    if (~a & 1):                                                #(D)
        if (b &1):                                              #(E)
            return bgcd(a >> 1, b)                              #(F)
        else:                                                   #(G)
            return bgcd(a >> 1, b >> 1) << 1                    #(H)
    if (~b & 1):                                                #(I)
        return bgcd(a, b >> 1)                                  #(J)
    if (a > b):                                                 #(K)
        return bgcd( (a-b) >> 1, b)                             #(L)
    return bgcd( (b-a) >> 1, a )                                #(M)

gcdval = bgcd(a, b)
print("\nBGCD: %d\n" % gcdval)
